enhancements/update/upgrade-without-registry.md

```diff
@@ -0,0 +1,989 @@
+
+Note that this process will by default work using whatever registry servers are
+configured in the cluster. When it isn't possible to use a registry server, the
+administrator of the cluster will explicitly configure the cluster version
+operator to use an upgrade bundle instead, using the new
+`spec.desiredUpdate.bundle.enabled: true` field:
+
+```yaml
+apiVersion: config.openshift.io/v1
+kind: ClusterVersion
+metadata:
+  name: version
+spec:
+  desiredUpdate:
+    bundle:
+      enabled: true
+```
+
+The technician that performs the upgrade will receive the USB stick containing
+the upgrade bundle and will plug it in one of the nodes of the cluster.
+
+The cluster version operator will be monitoring device events in all the nodes
+of the cluster in order to detect when such an USB stick (or any other kind of
+media containing an upgrade bundle) has been plugged. To control that a new
+`spec.desiredUpdate.bundle.monitor` field will be added to the `ClusterVersion`
+object. The default value will be `false` so that this monitoring will be
+disabled by default. When set to `true` the cluster version operator will create
+a new `bundle-monitor` daemon set that will perform the actual monitoring. When
+any of the pods in this daemon set detects a device containing a valid upgrade
+bundle it will update the status of the `ClusterVersion` object indicating that
+the bundle is available, via a new `status.desired.bundle.file` field.
+
+For example, if the administrator of the cluster wants to enable automatic
+detection of upgrade bundles she will add this to the `ClusterVersion` object:
+
+```yaml
+apiVersion: config.openshift.io/v1
+kind: ClusterVersion
+metadata:
+  name: version
+spec:
+  desiredUpdate:
+    bundle:
+      enabled: true
+      monitor: true
+```
+
+For situations where it isn't desirable to monitor device events, or it isn't
+possible to plug the USB stick or any other kind of media, we will also add a
+new `spec.desiredUpdate.bundle.file` field to explicitly indicate the location
+of the bundle. When this is used the value will be directly copied to the
+`status.desired.bundle.file` field, without creating the `bundle-monitor` daemon
+set.
+
+If the administrator wants to disable automatic detection of the upgrade bundle,
+and wants to manually copy the file to one of the cluster nodes she will do
+something like this:
+
+```yaml
+spec:
+  desiredUpdate:
+    bundle:
+      enabled: true
+      file: /root/upgrade-4.13.7-x86_64.tar
+```
+
+If the administrator wants to disable automatic detection, but wants to use a
+USB stick anyhow he will do something like this:
+
+```yaml
+spec:
+  desiredUpdate:
+    bundle:
+      enabled: true
+      file: /dev/sdb
+```
+
+When the `status.desired.bundle.file` field is populated the cluster version
+operator will start to replicate the bundle to the rest of the nodes of the
+cluster. To do so it will first disable auto-scaling to ensure that no new
```

Comment:
Eek.  That sounds like an exciting new fragile cross-operator connection.

I think instead we could inhibit new nodes joining the cluster (as distinct from new machines).  That'd be part of https://github.com/openshift/cluster-machine-approver

But...do we really need this?  We already today inherently have a race between upgrades and scaleup.  In the MCO today we serve the latest config if at least one node has upgraded successfully.

Hmm...wouldn't a lot of the concerns here in general be obviated if in this scenario we defaulted to cross-node image fetching?  IOW if at least one node in the cluster (e.g. control plane) has an image and we're in an offline upgrade scenario, default to fetching from it.

I suspect (particularly for the daemonset container images) it'd really benefit us to do peer-to-peer fetches by default across the board.

Reply:
The intent here is to prevent addition of new nodes after the images have been replicated to all the nodes and before the upgrade is started. If that happened then those new nodes wouldn't have the images and wouldn't be able to upgrade.